<?xml version="1.0"?>
<robot name="scara" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Constants for robot dimensions -->
  <xacro:property name="PI" value="3.1415926535897931"/>
  <xacro:property name="mass" value="1" /> <!-- arbitrary value for mass -->
  <xacro:property name="width" value="0.1" /> <!-- Square dimensions (widthxwidth) of beams -->
  <xacro:property name="height" value="1" /> <!-- Link height -->
  <xacro:property name="mass_pole" value="0.1" /> <!-- arbitrary value for mass -->
  <xacro:property name="width_pole" value="0.1" /> <!-- Square dimensions (widthxwidth) of beams -->
  <xacro:property name="height_pole" value="0.1" /> <!-- Link height -->
  <xacro:property name="prismatic_limit" value="1" /> <!-- Limit for prismatic joint -->
  <xacro:property name="damping_coeff" value="0.7" /> <!-- Damping coefficient for joints -->

  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <xacro:include filename="$(find rbe500_project)/urdf/scara.gazebo" />

  <!-- Used for fixing robot to Gazebo 'base' -->


  <link name="world"/>


  <link name="link1">
    <collision>
      <geometry>
        <box size="${width} ${width} ${height}"/>
      </geometry>
      <origin rpy="0 0 ${-PI}" xyz="0 0 ${height/2}"/>
    </collision>

    <visual>
      <geometry>
        <box size="${width} ${width} ${height}"/>
      </geometry>
      <origin rpy="0 0 ${-PI}" xyz="0 0 ${height/2}"/>
    </visual>

    <inertial>
      <origin xyz="0 0 ${-PI}" rpy="0 0 ${height/2}"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + height*height)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height*height + width*width)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + width*width)}"/>
    </inertial>
  </link>

  <joint name="theta1" type="fixed">
    <parent link="world"/>
    <child link="link1"/>
    <limit effort="1000.0" lower="${-PI}" upper="${PI}" velocity="0.5"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0 0 0"/>
    <dynamics damping="${damping_coeff}"/>
  </joint>

  <link name="link2">
    <collision>
      <geometry>
        <box size="${width} ${width} ${height}"/>
      </geometry>
      <origin rpy="0 ${PI/2} 0" xyz="${height/2} 0 0"/>
    </collision>

    <visual>
      <geometry>
        <box size="${width} ${width} ${height}"/>
      </geometry>
      <material name="my_color">
        <color rgb="255 135 255" alpha="1.0"/>
      </material>
      <origin rpy="0 ${PI/2} 0" xyz="${height/2} 0 0"/>
    </visual>

    <inertial>
      <origin rpy="0 ${PI/2} 0" xyz="${height/2} 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + height*height)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height*height + width*width)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + width*width)}"/>
    </inertial>
  </link>

  <joint name="bend" type="fixed">
    <parent link="link1"/>
    <child link="link2"/>
    <origin xyz="0 0 ${height}"/>
  </joint>

  <link name="link3">
    <collision>
      <geometry>
        <box size="${width} ${width} ${height}"/>
      </geometry>
      <origin rpy="0 ${PI/2} 0" xyz="${height/2} 0 0"/>
    </collision>
  
    <visual>
      <geometry>
        <box size="${width} ${width} ${height}"/>
      </geometry>
      <origin rpy="0 ${PI/2} 0" xyz="${height/2} 0 0"/>
    </visual>

    <inertial>
      <origin rpy="0 ${PI/2} 0" xyz="${height/2} 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + height*height)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height*height + width*width)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + width*width)}"/>
    </inertial>
  </link>

  <joint name="theta2" type="fixed">
    <parent link="link2"/>
    <child link="link3"/>
    <limit effort="1000.0" lower="${-PI}" upper="${PI}" velocity="0.5"/>
    <axis xyz="0 0 1"/>
    <origin xyz="${height} 0 0"/>
    <dynamics damping="${damping_coeff}"/>
  </joint>

  <link name="gripper_pole">
    <collision>
      <geometry>
        <box size="${width_pole} ${width_pole} ${height_pole}"/>
      </geometry>
      <origin rpy="0 0 0 " xyz="0 0 0"/>
    </collision>

    <visual>
      <geometry>
        <box size="${width_pole} ${width_pole} ${height_pole}"/>
      </geometry>
      <origin rpy="0 0 0 " xyz="0 0 0"/>
    </visual>

    <inertial>
      <origin rpy="0 0 0 " xyz="0 0 0"/>
      <mass value="${mass_pole}"/>
      <inertia
	  ixx="${mass / 12.0 * (width_pole*width_pole + height_pole*height_pole)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height_pole*height_pole + width_pole*width_pole)}" iyz="0.0"
	  izz="${mass / 12.0 * (width_pole*width_pole + width_pole*width_pole)}"/>
    </inertial>
  </link>

  <joint name="d3" type="prismatic">
    <parent link="link3"/>
    <child link="gripper_pole"/>
    <limit effort="1000.0" lower="${-prismatic_limit}" upper="${prismatic_limit}" velocity="0.5"/>
    <axis xyz="0 0 1"/>
    <origin rpy="${PI} 0 0" xyz="${height} 0 0"/>
    <dynamics damping="${damping_coeff}"/>
  </joint>

  <gazebo reference="gripper_pole">
    <turnGravityOff>true</turnGravityOff>
  </gazebo>

</robot>